addons/mail/tests/test_mail_canned_response.py

```diff
@@ -0,0 +1,78 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo.addons.mail.tests.common import MailCommon
+from odoo.exceptions import AccessError
+from odoo.tests.common import tagged, HttpCase
+from odoo import Command
+
+
+@tagged("mail_canned_response", "post_install", "-at_install")
+class TestCannedResponse(MailCommon, HttpCase):
+
+    def test_user_canned_response(self):
```

Comment:
as you are globally always doing tow main tests (one calling mail/data, one for write or unlink access), it would be better write one unit test per feature (one for mail/data, one for write and unlink). In that test, have inputs and expected outputs, and do tests in a loop using ``subTest`` . There are various examples in test_mail notably if you want to check how to write them.

It will allow to have less diff, less code, and improve coverage readability: we know what use cases are covered and it would be easy to add a new one (e.g. add portal just in intput / expected instead of having to rewrite the whole test).

Reply:
I don't like this way of doing testing, but whatever, I changed the code.

Reply:
Wasn't it changed in a way that just reduced what is tested?

The idea of doing it in a loop would be to cover all cases, in a similar way than `TestDiscussChannelAccess` (hopefully much simpler).

I didn't know about `with self.subTest` but it doesn't look like it brings any facility to write the test, just a better output in case of error?

Reply:
> Wasn't it changed in a way that just reduced what is tested?
> 
> The idea of doing it in a loop would be to cover all cases, in a similar way than `TestDiscussChannelAccess` (hopefully much simpler).
> 
> I didn't know about `with self.subTest` but it doesn't look like it brings any facility to write the test, just a better output in case of error?

subTest skips what is under its scope when failing. So you don't have to jump from error to error, they are all listed. It allows to make a single test (testing a feature) without having to fix issue one by one. So it is usefull when
  * developing a feature (test driven dev): write tests then make them green;
  * refactoring: you see all failing sub tests
  * making tests with several steps: you can move forward in the test even if a specific check in a test failed

All in all, purpose is to avoid the setup and writing of redundant tests when the core is the same. You just loop on input / expected output.